commit 469b1a7280c58a40d718686f26c0ae0e2f472436 (refs/changes/51/551/4)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-09-30T00:19:49+00:00 (3 years ago)
    
    joyent/node-cueball#30 hang when max == count and everything is dead
    Reviewed by: David Pacheco <dap@joyent.com>

diff --git a/lib/utils.js b/lib/utils.js
index 17588ba..74a8133 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -194,15 +194,15 @@ function planRebalance(inSpares, dead, target, max, singleton) {
 		var count = done + replacements - i;
 		var empties = keys.filter(function (kk) {
 			if (singleton) {
-				return (dead[kk] !== true && (
-				    wantedSpares[kk] === undefined ||
-				    wantedSpares[kk] === 0));
+				return (dead[kk] !== true &&
+				    wantedSpares[kk] === undefined);
 			} else {
 				return (dead[kk] !== true ||
-				    wantedSpares[kk] === undefined ||
-				    wantedSpares[kk] === 0);
+				    wantedSpares[kk] === undefined);
 			}
 		});
+
+		/* We have room for both this and a replacement. */
 		if (count + 1 <= max) {
 			if (wantedSpares[k] === 0) {
 				wantedSpares[k] = 1;
@@ -211,8 +211,23 @@ function planRebalance(inSpares, dead, target, max, singleton) {
 			if (empties.length > 0)
 				++replacements;
 
+		/*
+		 * We only have room for one, but there are other candidates
+		 * that might actually be up. Use one of them instead.
+		 */
 		} else if (count <= max && empties.length > 0) {
 			++replacements;
+
+		/* Only room for one, everything looks dead. Use us. */
+		} else if (count <= max) {
+			if (wantedSpares[k] === 0) {
+				wantedSpares[k] = 1;
+				++done;
+			}
+
+		/* Already met our max socket cap. Give up now. */
+		} else {
+			break;
 		}
 	}
 
diff --git a/test/utils.test.js b/test/utils.test.js
index 16950f1..ea21933 100644
--- a/test/utils.test.js
+++ b/test/utils.test.js
@@ -263,3 +263,23 @@ mod_tape.test('rebalance: dead, backend starvation', function (t) {
 	t.deepEqual(plan.add, ['b2', 'b2', 'b2']);
 	t.end();
 });
+
+mod_tape.test('bug #30', function (t) {
+	var spares = {
+		'16uN6JsJFild9cHyl2+LSyRHmNc=': ['c1'],
+		'c7QG0UOYCpm6m/hYUX0jBenbM70=': ['c2'],
+		'ashWtupYHh1QH33UP/T2+6hvi8c=': [],
+		'4QMg6SChOmtF8s6lfK32lLoKUFs=': []
+	};
+	var dead = {
+		'c7QG0UOYCpm6m/hYUX0jBenbM70=': true,
+		'16uN6JsJFild9cHyl2+LSyRHmNc=': true,
+		'4QMg6SChOmtF8s6lfK32lLoKUFs=': true,
+		'ashWtupYHh1QH33UP/T2+6hvi8c=': true
+	};
+	var plan = mod_utils.planRebalance(spares, dead, 3, 4);
+	t.deepEqual(plan.remove, []);
+	t.deepEqual(plan.add, [
+	    'ashWtupYHh1QH33UP/T2+6hvi8c=', '4QMg6SChOmtF8s6lfK32lLoKUFs=']);
+	t.end();
+});
